exploration/0002-expression-attributes.md

```diff
@@ -0,0 +1,126 @@
+or other similar factors that influence the formatted result.
+
+However, this single bag of options is not appropriate in all cases,
+in particular for attributes that pertain to the expression as a selector or a placeholder.
+For example, many of the [XLIFF 2 inline element] attributes don't really make sense as function options.
+
+[XLIFF 2 inline element]: http://docs.oasis-open.org/xliff/xliff-core/v2.1/os/xliff-core-v2.1-os.html#inlineelements
+
+## Use-Cases
+
+At least the following expression attributes should be considered:
+
+- Attributes with a formatting runtime impact:
+
+  - `fallback` — A value to use instead of the default fallback,
+    should the expression's primary formatting fail in some way.
+  - `locale` — An override for the locale used to format the expression.
+    Should be expressed as a non-empty sequence of BCP 47 language codes.
+  - `dir` — An override for the LTR/RTL/auto directionality of the expression.
+
+- Attributes relevant for translators, tools, and other message operations,
+  but with no runtime impact:
+
+  - `example` — A literal value representing
+    what the expression's formatted value will look like.
+  - `note` — A comment on the expression for translators.
+  - `translate` — A boolean `yes`/`no` indicator communicating to translators
+    whether the expression should or should not be localised.
+  - `canCopy`, `canDelete`, `canOverlap`, `canReorder`, etc. — Flags supported by
+    XLIFF 2 inline elements
+
+## Requirements
+
+Attributes should be assignable to un-annotated expressions.
+
+Attributes should be distinct from function options.
+
+Common attributes should be definable by the specification.
+
+Attributes should be user-definable.
+
+Attributes with a runtime impact should not be considered comments.
+
+Attributes should relate to a specific expression.
+
+Multiple attributes should be assignable to a single expression.
+
+Attributes should be assignable to all expressions, not just placeholders.
+
+## Constraints
+
+If supported by new syntax,
+the syntax should be easy to parse by both humans and machines.
+
+If supported by new syntax,
+the reserved/private-use rules will need to be adjusted.
+
+## Proposed Design
+
+Add support for option-like `@key=value` attribute pairs at the end of any expression.
+
+To distinguish expression attributes from options,
+require `@` as a prefix for each attribute asignment.
+Examples: `@translate=yes` and `@locale=$exprLocale`.
+
+Define the meaning and supported values of some expression attributes in the specification,
+including at least `@dir` and `@locale`.
+
+To support later extension of the specified set of attributes while allowing user extensibility,
+require custom attribute names to include a U+002D Hyphen-Minus `-`.
+Examples: `@can-copy=no`, `@note-link=|https://...|`.
+
+Allow expression attributes to influence the formatting context,
+but do not directly pass them to user-defined functions.
```

Comment:
I would tighten this up. I would imagine that we'll add to the spec text similar to:

> An implementation MUST provide functions the ability to query the set of attributes that apply to the calling expression.

(I didn't say "an API" on purpose)

Reply:
What you're proposing is actually the opposite of what I'm proposing. The formatting context as [currently specified](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#formatting-context) is an explicit set of fields, extensible by an implementation but _**not**_ custom functions:
> Implementations MAY include additional fields in their formatting context.

So the proposal here would allow an implementation to modify the formatting context as it's applied to an expression, but it would not allow for a custom function definition "to query the set of attributes".

If it makes sense for an attribute to influence the behaviour of a custom function, and that attribute isn't an explicit part of the implementation's formatting context, why should its value be communicated via an expression attribute rather than a function option?

Reply:
A formatting context would be "a way to query the set of attributes"? I like your wording:

> Implementations MAY include additional fields in their formatting context.

Although my question would be: how do I ensure that my custom annotation reaches my function's runtime via the context?

Reply:
That wording is from our current formatting spec, actually.

> how do I ensure that my custom annotation reaches my function's runtime via the context?

By using a function option to convey that information, rather than an expression attribute.

Reply:
My thinking here goes:

* We are defining a feature "expression attributes"
* We allow implementation or user defined attributes
* How do these attributes reach the runtime?

The answer cannot be function options or I would have used a function option and not an expression attribute. 

Apparently the answer is: the attributes appear in the formatting context and we can just say that?

Reply:
In that situation, why not use a function option? If it's something affecting the behaviour of a single function and its value can be defined in the expression, _why not_ use a function option for it?

> Apparently the answer is: the attributes appear in the formatting context and we can just say that?

We have not defined the shape of the formatting context with any specificity, and I don't think we should -- it's really an implementation detail. If an implementation wants to make its formatting context extensible in some way and for a custom expression attribute to affect it, I think that's very much an implementation detail and outside the scope of this spec.

Reply:
We don't have to define the shape of the formatting context with any specificity in order to say that implementations must provide access to any attribute values in a message's context. We already have at least one contextual value (the locale) that we require the implementation to ensure. 

So I agree that we don't want to define exactly how a function gets access to the context. We might not even require that there be a way for it to be obtained. But we can still require that any contextual mechanism (using whatever shape) provides access to the attribute.

If we don't follow my thinking through, wouldn't we be introducing a feature that doesn't work in at least some implementations?

Reply:
To rephrase and reiterate my own position on this and the currently proposed phrase, I do not think that functions should have access to the expression attributes or their values.

I do not think it would be a good idea for implementations to be required to provide access to any attribute values in a message's context. As we already have function options, I do not think that expression attributes should be duplicating their functionality. Allowing custom functions to access expression attributes would do exactly that.

Message formatting happens in some context that has properties such as the current locale, timezone, etc. that have an effect on the results. In the spec, we refer to this as the "formatting context", but do not exhaustively define its shape or properties. This is implementation-defined, and implementations may internally allow for its extensibility.

The general use case I have in mind for expression attributes that have a formatting impact is to provide a way to locally override formatting context values or error handling behaviour.

What is the use case for allowing custom functions to access custom attributes?